File: Sources/WickSync/JournalDayMerge.swift
import Foundation

/// Outcome of merging two device versions of the same journal day.
public struct JournalDayMergeResult: Equatable {
    /// The union entry both devices should converge to.
    public var merged: JournalEntry
    /// Same-item-id conflicts whose contents differed — the losing copies.
    /// Surfaced to the user (and archived remotely) so nothing vanishes silently.
    public var losingItems: [JournalItem]
    /// The discarded title when both sides had different non-empty titles.
    public var losingTitle: String?
}

/// Merges two versions of one day (same `dayKey`) that diverged on two devices.
///
/// Rules (deterministic so all devices converge to the same result):
/// - Entry identity (id/date/dayKey) comes from the older entry by `createdAt`.
/// - Items are unioned by item id — the data model's one-day-many-items shape
///   means concurrent edits usually touch different items and merge cleanly.
/// - Same item id with different contents: the side with the newer
///   `entry.updatedAt` wins; the loser is preserved in `losingItems`.
/// - Title: identical or one-sided-empty resolves trivially; two different
///   non-empty titles go to the newer side, loser recorded in `losingTitle`.
public enum JournalDayMerge {
    public static func merge(local: JournalEntry, remote: JournalEntry) -> JournalDayMergeResult {
        let localIsBase = local.createdAt <= remote.createdAt
        let base = localIsBase ? local : remote
        let other = localIsBase ? remote : local
        // Newer side wins per-item conflicts; exact ties go to local (the rev
        // race afterwards converges devices even if they pick differently).
        let localWinsConflicts = local.updatedAt >= remote.updatedAt

        var itemsByID: [UUID: JournalItem] = [:]
        var order: [UUID] = []
        for item in base.items {
            itemsByID[item.id] = item
            order.append(item.id)
        }

        var losingItems: [JournalItem] = []
        for item in other.items {
            guard let existing = itemsByID[item.id] else {
                itemsByID[item.id] = item
                order.append(item.id)
                continue
            }
            guard existing != item else { continue }

            // `existing` came from base, `item` from other — map back to
            // local/remote to apply the newer-updatedAt rule.
            let localCopy = localIsBase ? existing : item
            let remoteCopy = localIsBase ? item : existing
            let winner = localWinsConflicts ? localCopy : remoteCopy
            let loser = localWinsConflicts ? remoteCopy : localCopy
            itemsByID[item.id] = winner
            if !loser.isEmpty {
                losingItems.append(loser)
            }
        }

        var items = order.compactMap { itemsByID[$0] }
        // Placeholder empty items add nothing once real content exists.
        let nonEmpty = items.filter { !$0.isEmpty }
        if !nonEmpty.isEmpty {
            items = nonEmpty
        }
        if items.isEmpty {
            items = [JournalItem()]
        }

        let localTitle = local.title.trimmingCharacters(in: .whitespacesAndNewlines)
        let remoteTitle = remote.title.trimmingCharacters(in: .whitespacesAndNewlines)
        var title = base.title
        var losingTitle: String?
        if localTitle != remoteTitle {
            if localTitle.isEmpty {
                title = remote.title
            } else if remoteTitle.isEmpty {
                title = local.title
            } else {
                title = localWinsConflicts ? local.title : remote.title
                losingTitle = localWinsConflicts ? remote.title : local.title
            }
        }

        var merged = base
        merged.title = title
        merged.items = items
        merged.createdAt = min(local.createdAt, remote.createdAt)
        merged.updatedAt = max(local.updatedAt, remote.updatedAt)

        return JournalDayMergeResult(merged: merged, losingItems: losingItems, losingTitle: losingTitle)
    }
}

File: linux/src/sync/JournalDayMerge.cpp
#include "JournalDayMerge.h"

#include <algorithm>
#include <map>
#include <stdexcept>

namespace wick {

JournalEntryMergeResult JournalEntryMerge::merge(const JournalEntry& local, const JournalEntry& remote) {
    if (!(local.id == remote.id)) {
        throw std::logic_error("Only versions of the same entry may merge");
    }
    const bool localIsBase = local.createdAt <= remote.createdAt;
    const JournalEntry& base = localIsBase ? local : remote;
    const JournalEntry& other = localIsBase ? remote : local;
    const bool localWinsConflicts = local.updatedAt >= remote.updatedAt;

    std::map<Uuid, JournalItem> itemsByID;
    std::vector<Uuid> order;
    for (const auto& item : base.items) {
        itemsByID[item.id] = item;
        order.push_back(item.id);
    }

    std::vector<JournalItem> losingItems;
    for (const auto& item : other.items) {
        auto it = itemsByID.find(item.id);
        if (it == itemsByID.end()) {
            itemsByID[item.id] = item;
            order.push_back(item.id);
            continue;
        }
        if (it->second == item) continue;

        const JournalItem existing = it->second;
        const JournalItem localCopy = localIsBase ? existing : item;
        const JournalItem remoteCopy = localIsBase ? item : existing;
        const JournalItem winner = localWinsConflicts ? localCopy : remoteCopy;
        const JournalItem loser = localWinsConflicts ? remoteCopy : localCopy;
        itemsByID[item.id] = winner;
        if (!loser.isEmpty()) losingItems.push_back(loser);
    }

    std::vector<JournalItem> items;
    for (const auto& id : order) {
        auto it = itemsByID.find(id);
        if (it != itemsByID.end()) items.push_back(it->second);
    }
    std::vector<JournalItem> nonEmpty;
    for (const auto& item : items) {
        if (!item.isEmpty()) nonEmpty.push_back(item);
    }
    if (!nonEmpty.empty()) items = std::move(nonEmpty);
    if (items.empty()) items.push_back(JournalItem{Uuid::generate(), "", "", {}, std::nullopt});

    const std::string localTitle = trimCopy(local.title);
    const std::string remoteTitle = trimCopy(remote.title);
    std::string title = base.title;
    std::optional<std::string> losingTitle;
    if (localTitle != remoteTitle) {
        if (localTitle.empty()) {
            title = remote.title;
        } else if (remoteTitle.empty()) {
            title = local.title;
        } else {
            title = localWinsConflicts ? local.title : remote.title;
            losingTitle = localWinsConflicts ? remote.title : local.title;
        }
    }

    JournalEntry merged = base;
    merged.id = local.id;
    merged.date = localWinsConflicts ? local.date : remote.date;
    merged.title = title;
    merged.items = std::move(items);
    merged.createdAt = std::min(local.createdAt, remote.createdAt);
    merged.updatedAt = std::max(local.updatedAt, remote.updatedAt);

    return JournalEntryMergeResult{std::move(merged), std::move(losingItems), std::move(losingTitle)};
}

} // namespace wick
